[REVIEW] Consolidate functionality in NestedHostColumnVector and HostColumnVector [skip-ci]

kuhushukla opened the pull request:
Fixes #5905 .

I have marked it as WIP since this does not totally obliterate the developer from having the knowledge of a new class `BaseHostColumnVector` but it does consolidate the offHeap state on the host side and keeps  `refCount` in the `HostColumnVector`. Also, I _think_ I free memory correctly here per the revision but I need to test a bit more. All unit tests show no additional leaks and memory on the GPU, while these tests run, seems ok. 

Additionally, we could do a different approach here with composition to totally hide the `BaseHostColumnVector` but I could use some reviews on that idea and this PR first.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

kuhushukla commented:
@revans2 appreciate any thoughts on this change and the what modifications I can make to make this better.

revans2 commented on java/src/main/java/ai/rapids/cudf/BaseHostColumnVector.java:
```diff
@@ -0,0 +1,267 @@
+import java.util.List;
+import java.util.Optional;
+
+public class BaseHostColumnVector implements AutoCloseable {
```
nit: could we at least have a simple java doc description of the class?

It would be nice to have some simple javadocs for all of the APIs added here.  I know they didn't exist before.

revans2 commented on java/src/main/java/ai/rapids/cudf/BaseHostColumnVector.java:
```diff
@@ -0,0 +1,267 @@
+    return type;
+  }
+
+  long getNullCount() {
```
Before this returned the `Optional<Long>` is there a reason for the change?  If we always have the null count then we should just remove the Optional everywhere.  Is this a breaking change?

kuhushukla replied:
If we keep it Optional<Long> as return type, the `getNullCount` method conflicts with the oversion from before in `HostColumnVector`

I dont think it is a breaking change - I will test it with the plugin but I think we can keep this as `long getNullCount()`

revans2 replied:
Sounds like we need to resolve this at some point for consistency sake.  I don't think it is a big deal right now.

revans2 commented on java/src/main/java/ai/rapids/cudf/BaseHostColumnVector.java:
```diff
@@ -0,0 +1,267 @@
+    return children;
+  }
+
+  public HostMemoryBuffer getOffsets() {
```
Nit could you put offsets next to data and validity.  They feel like they should go together.

revans2 commented on java/src/main/java/ai/rapids/cudf/BaseHostColumnVector.java:
```diff
@@ -0,0 +1,267 @@
+    return nullCount.get();
+  }
+
+  List<BaseHostColumnVector> getNestedChildren() {
```
nit: I'm not a fan of exposing `BaseHostColumnVector` as a name in the public API. It is playing a dual role as both the `Base` for `HostColumnVector` and as the children of a `HostColumnVector`.  Perhaps `HostColumnVectorCore` or `CoreHostColumnVector` might be better?

@jlowe can you think of a better name for this? or if you have a similar concern?

revans2 reviewed (approved):
Really just some nits

revans2 commented:
@kuhushukla this appears to be a breaking change.  Do you have a patch to https://github.com/NVIDIA/spark-rapids/ that adjusts to these changes?

kuhushukla commented:
@revans2 I am going to rebase this and address the nit comments as part of that once #6171 is in. Thanks for the review.

kuhushukla commented:
Going to rebase and refurbish it today.

kuhushukla commented:
> Additionally - will put the plugin side PR up soon.

This is not required anymore per latest PR update. I have rebased and refactored this a bit and could use another round of review @revans2. Appreciate the reviews.
